lambda_src/drivers/destination_s3.py

```diff
@@ -89,35 +87,43 @@ def initialize(destination: Text, batch_id: Text):
     # Regex captures characters after and including the rightmost '/' in a path,
     # which are then replaced with a '/', e.g. '/a/b/c' -> '/a/b/'
     prefix_folder = re.sub(r'/[^/]*$', '/', prefix) if '/' in prefix else ''
-
+    prefix_folder = re.sub(r'/{[^}].*', '/', prefix_folder)
     if prefix_folder:
         write_to_s3(bucket, prefix_folder, content)
 
 
 def write(
     destination: Text,
     batch_id: Text,
-    datum: Union[Dict, List],
+    datum: Union[Dict, List, bytes],
     row_index: int,
 ) -> Dict[Text, Any]:
     bucket, prefix = parse_destination_uri(destination)
-    encoded_datum = (
-        '\n'.join(json.dumps(d) for d in datum)
-        if isinstance(datum, list)
-        else json.dumps(datum, default=str)
-    )
+    if isinstance(datum, bytes):
+        encoded_datum = datum
+        encoded_datum_hash = sha256(encoded_datum).hexdigest()
+        prefixed_filename = prefix.format(hash=encoded_datum_hash)
+    else:
+        encoded_datum = (
+            '\n'.join(json.dumps(d) for d in datum)  # type: ignore
```

Comment:
Since `encoded_datum` can be of type `str` or `bytes`, mypy produces the following error:

```
error: Incompatible types in assignment (expression has type "str", variable has type "bytes")
```

If the type for `encoded_datum` is set to `Union[bytes, str, AnyStr]`, mypy produces this error:

```
error: Value of type variable "AnyStr" of "write_to_s3" cannot be "Union[bytes, str]"
```

This is seems to be an open bug: https://github.com/python/mypy/issues/1533

So best to just ignore the type for now? @sfc-gh-afedorov

Reply:
> Long story short, If we're going to use mypy it's best not ignore its errors and add an explicit type declaration at the top of the function, like encoded_datum: Union[bytes, str].

True, I'd tried forward declaring the type actually, but it ended up with the AnyStr error. 

> In this case, tho, I think it's best cast the right hand value on line 107 to bytes so that it's aligned with what the boto3 function is documented to receive, computes the hash in both cases, and removes the need for the if conditional and response variable below

Oo that works 👍